File: model_dev_and_vld/smi_r/R/constr_ols.R
#ssq function
ssq <- function(beta.coef, X, y, weights) {   
      resid <- c((t(y) - as.matrix(t(beta.coef))%*%t(X)))
      resid.w <- sum(weights * resid^2)
return(resid.w)
}

#' Constrained OLS Regrssion
#'@import dplyr
#'@export
constr.ols <- function(db, target, predictors, lower, upper, weights) {
      warn.init <- getOption("warn") 
      options(warn = -1)
      on.exit(options(warn = warn.init))

      X <- cbind(1, db[, predictors, drop = FALSE])
      y <- db[, target]
      lower <- c(-Inf, lower)
      upper <- c(Inf, upper)
      cc <- complete.cases(cbind(y, X))
      #initial values
      db.i <- data.frame(y, X)
      frm <- paste0("y ~ ", paste(names(X)[-1], collapse = " + "))
      start.vls <- unname(coef(lm(formula = frm,
                                  data = db.i)))
      opt <- optim(par = start.vls,
                   fn = ssq,
                   X = X[cc, ],
                   y = y[cc],
                   weights = weights[cc],
                   lower = lower,
                   upper = upper,
                   method = "L-BFGS-B")
      #betas
      beta.opt <- opt$par
      names(beta.opt) <- c("(Intercept)", predictors) 
      #model fit
      pred.opt <- c(as.matrix(t(beta.opt))%*%t(X))
      #r-squared
      r.squared <- cor(x = y, 
                       y = pred.opt,
                       use = "complete.obs")^2
      #loglikelihood
      n <- sum(cc)
      k <- sum(abs(beta.opt[-1]) > 1/1e6) + 2 #betas + intercept + 1 for sigma.hat
      rss <- sum((y - pred.opt)^2, na.rm = TRUE)
      sigma.hat <- rss / n
      logL <- -n/2 * (log(2*pi) + log(sigma.hat) + 1)
      #aic
      aic <- 2 * k - 2 * logL
      #bic
      bic <- log(n) * k - 2 * logL
      #summary
      res <- list(beta = beta.opt, pred = pred.opt, r.squared = r.squared, aic = aic, bic = bic)    
return(res)
}

#' Constrained Models Estimation
#'@import dplyr
#'@export
model.est <- function(gr.c, ps, db, target, weights) {
      warn.init <- getOption("warn") 
      options(warn = -1)
      on.exit(options(warn = warn.init))

      smi.r <- nrow(gr.c)
      res <- vector("list", smi.r) 
      pred <- vector("list", smi.r) 
      
      for   (i in 1:smi.r) {
            predi.i <- unname(c(gr.c[i, ], recursive = TRUE))
            predi.i <- predi.i[!is.na(predi.i)]
            #predictor name
            pred.n <- sub("_lag[1-4]$", "", predi.i)
            #lower and upper bounds
            lower <- c()
            upper <- c()
            for   (j in 1:length(pred.n)) {
                  ps.j <- ps[pred.n[j]]
                  if    (ps.j == "+") {
                        lower <- c(lower, 0)
                        upper <- c(upper, Inf)
                        } 
                  if    (ps.j == "-") {
                        lower <- c(lower, -Inf)
                        upper <- c(upper, 0)
                        }
                  if    (ps.j == "+/-") {
                        lower <- c(lower, -Inf)
                        upper <- c(upper, Inf)
                        }
                  }
            #constrained ols      
            res.i <- constr.ols(db = db, 
                                target = target, 
                                predictors = predi.i, 
                                lower = lower, 
                                upper = upper, 
                                weights = weights)
            pred[[i]] <- res.i$pred
            #summary
            res.i <- data.frame(model.id = paste0("Model_", i), 
                                coefficient = names(res.i$beta), 
                                estimate = unname(res.i$beta), 
                                r.squared = res.i$r.squared,
                                aic = res.i$aic,
                                bic = res.i$bic,
                                zero.coeff = any(round(unname(res.i$beta), 6) == 0))
            res.i$predictor <- sub("_lag[1-4]$", "", res.i$coefficient)
            res.i$lag <- get.lag(x = res.i$coefficient)
            #storing
            res[[i]] <- res.i
            }	 
            res <- bind_rows(res)
return(list(models = res, pred = pred))
}

#' Bivariate Filtering
#'@import dplyr
#'@export
bf <- function(db, target, group, lower, upper, weights) {
      warn.init <- getOption("warn") 
      options(warn = -1)
      on.exit(options(warn = warn.init))

      pl <- length(group)
      res <- vector("list", pl)
      for   (i in 1:pl) {
            predictor.i <- group[i]
            res.i <- constr.ols (db = db, 
                                 target = target, 
                                 predictors = predictor.i , 
                                 lower = lower, 
                                 upper = upper, 
                                 weights = weights)
            res.i <- data.frame(predictor = predictor.i,
                                coefficient = names(res.i$beta), 
                                estimate = unname(res.i$beta), 
                                r.squared = res.i$r.squared)
            res[[i]] <- res.i
            }
      res <- do.call("rbind", res)
      if    (all(is.na(res$r.squared))) {
            pred.select <- NULL
            } else {
            pred.select <- unique(res$predictor[which.max(res$r.squared)])
            }
      res <- list(summary = res, selected = pred.select)
return(res)
}

File: model_dev_and_vld/smi_py/smi/constr_ols.py
import pandas as pd
import numpy as np
from scipy.optimize import minimize

#ssq function
def ssq(beta_coef, X, y, weights):
    resid = y - X @ beta_coef
    resid_w = 0.5 * np.sum(weights * resid ** 2)
    return resid_w


#constrained ols regression
def constr_ols(db, target, predictors, lower, upper, weights):
    X = np.column_stack([np.ones(len(db)), db[predictors].values])
    y = db[target].values
    lower = np.concatenate(([-np.inf], lower))
    upper = np.concatenate(([np.inf], upper))
    cc = np.isfinite(np.column_stack([y, X])).all(axis = 1)

    #initial values
    db_i = pd.DataFrame(np.column_stack([y, X]),
                        columns = ["y"] + list(range(X.shape[1])))
    db_i = db_i[cc]
    y_lm = db_i["y"].values
    X_lm = db_i.iloc[:, 1:].values
    beta_start = np.linalg.lstsq(X_lm, y_lm, rcond = None)[0]
    #constrained ols
    opt = minimize(fun = ssq,
                   x0 = beta_start,
                   args = (X[cc, :], y[cc], weights[cc]),
                   bounds = list(zip(lower, upper)),
                   method="L-BFGS-B")
    #betas
    beta_opt = opt.x
    beta_names = ["(Intercept)"] + predictors
    #model fit
    pred_opt = X @ beta_opt
    #r-squared
    r_squared = np.corrcoef(y, pred_opt)[0, 1] ** 2
    #loglikelihood
    n = np.sum(cc)
    k = np.sum(np.abs(beta_opt[1:]) > 1/1e6) + 2
    rss = np.nansum((y - pred_opt)**2)
    sigma_hat = rss / n
    logL = -n/2 * (np.log(2*np.pi) + np.log(sigma_hat) + 1)
    #aic
    aic = 2 * k - 2 * logL
    #bic
    bic = np.log(n) * k - 2 * logL
    #summary
    res = {"beta": dict(zip(beta_names, beta_opt)),
           "pred": pred_opt,
           "r.squared": r_squared,
           "aic": aic,
           "bic": bic}

    return res


#constrained models estimation
def model_est(gr_c, ps, db, target, weights):
    smi_r = len(gr_c)
    res = [None] * smi_r
    pred = [None] * smi_r

    for i in range(smi_r):
        predi_i = list(gr_c[i])

        #predictor name
        pred_n = pd.Series(predi_i).str.replace(r"_lag[1-4]$", "", regex = True).values

        #lower and upper bounds
        lower = []
        upper = []
        for j in range(len(pred_n)):
            ps_j = ps[pred_n[j]]
            if ps_j == "+":
                lower.append(0)
                upper.append(np.inf)
            if ps_j == "-":
                lower.append(-np.inf)
                upper.append(0)
            if ps_j == "+/-":
                lower.append(-np.inf)
                upper.append(np.inf)

        #constrained ols
        res_i = constr_ols(db = db,
                           target = target,
                           predictors = predi_i,
                           lower = np.array(lower),
                           upper = np.array(upper),
                           weights = weights)

        pred[i] = res_i["pred"]

        #summary
        res_df = pd.DataFrame({"model.id": f"Model_{i+1}",
                               "coefficient": list(res_i["beta"].keys()),
                               "estimate": list(res_i["beta"].values()),
                               "r.squared": res_i["r.squared"],
                               "aic": res_i["aic"],
                               "bic": res_i["bic"],
                                "zero.coeff": any(np.round(list(res_i["beta"].values()), 6) == 0)})

        res_df["predictor"] = res_df["coefficient"].str.replace(r"_lag[1-4]$", "", regex = True)
        res_df["lag"] = get_lag(res_df["coefficient"].values)

        res[i] = res_df

    res = pd.concat(res, ignore_index=True)

    return {"models": res, "pred": pred}
